File: examples/online-boutique/mobile/lib/api/checkout.dart
// ignore_for_file: unused_import
library checkout;

import 'dart:convert';
import 'dart:typed_data';
import 'ftl_client.dart';
import 'builtin.dart' as builtin;
import 'cart.dart' as cart;
import 'currency.dart' as currency;
import 'payment.dart' as payment;
import 'productcatalog.dart' as productcatalog;
import 'shipping.dart' as shipping;


class ErrorResponse{
  String message;

  ErrorResponse({  required this.message,  });

  Map<String, dynamic> toJson() {
    return {
      'message': ((dynamic v) => v)(message),
    };
  }

  factory ErrorResponse.fromJson(Map<String, dynamic> map) {
    return ErrorResponse(
      message: ((dynamic v) => v)(map['message']), 
    );
  }
}

class Order{
  String id;
  String shippingTrackingId;
  currency.Money shippingCost;
  shipping.Address shippingAddress;
  List<OrderItem> items;

  Order({  required this.id,  required this.shippingTrackingId,  required this.shippingCost,  required this.shippingAddress,  required this.items,  });

  Map<String, dynamic> toJson() {
    return {
      'id': ((dynamic v) => v)(id),
      'shippingTrackingId': ((dynamic v) => v)(shippingTrackingId),
      'shippingCost': ((dynamic v) => v.toJson())(shippingCost),
      'shippingAddress': ((dynamic v) => v.toJson())(shippingAddress),
      'items': ((dynamic v) => v.map((v) => v.toJson()).cast<OrderItem>().toList())(items),
    };
  }

  factory Order.fromJson(Map<String, dynamic> map) {
    return Order(
      id: ((dynamic v) => v)(map['id']), shippingTrackingId: ((dynamic v) => v)(map['shippingTrackingId']), shippingCost: ((dynamic v) => currency.Money.fromJson(v))(map['shippingCost']), shippingAddress: ((dynamic v) => shipping.Address.fromJson(v))(map['shippingAddress']), items: ((dynamic v) => v.map((v) => OrderItem.fromJson(v)).cast<OrderItem>().toList())(map['items']), 
    );
  }
}

class OrderItem{
  cart.Item item;
  currency.Money cost;

  OrderItem({  required this.item,  required this.cost,  });

  Map<String, dynamic> toJson() {
    return {
      'item': ((dynamic v) => v.toJson())(item),
      'cost': ((dynamic v) => v.toJson())(cost),
    };
  }

  factory OrderItem.fromJson(Map<String, dynamic> map) {
    return OrderItem(
      item: ((dynamic v) => cart.Item.fromJson(v))(map['item']), cost: ((dynamic v) => currency.Money.fromJson(v))(map['cost']), 
    );
  }
}

class PlaceOrderRequest{
  String userId;
  String userCurrency;
  shipping.Address address;
  String email;
  payment.CreditCardInfo creditCard;

  PlaceOrderRequest({  required this.userId,  required this.userCurrency,  required this.address,  required this.email,  required this.creditCard,  });

  Map<String, dynamic> toJson() {
    return {
      'userId': ((dynamic v) => v)(userId),
      'userCurrency': ((dynamic v) => v)(userCurrency),
      'address': ((dynamic v) => v.toJson())(address),
      'email': ((dynamic v) => v)(email),
      'creditCard': ((dynamic v) => v.toJson())(creditCard),
    };
  }

  factory PlaceOrderRequest.fromJson(Map<String, dynamic> map) {
    return PlaceOrderRequest(
      userId: ((dynamic v) => v)(map['userId']), userCurrency: ((dynamic v) => v)(map['userCurrency']), address: ((dynamic v) => shipping.Address.fromJson(v))(map['address']), email: ((dynamic v) => v)(map['email']), creditCard: ((dynamic v) => payment.CreditCardInfo.fromJson(v))(map['creditCard']), 
    );
  }
}


class CheckoutClient {
  final FTLHttpClient ftlClient;

  CheckoutClient({required this.ftlClient});


  Future<Order> placeOrder(
    PlaceOrderRequest request, { 
    Map<String, String>? headers,
  }) async {
    final response = await ftlClient.post('/checkout/userId', request: request.toJson());
    if (response.statusCode == 200) {
      final body = json.decode(utf8.decode(response.bodyBytes));
      return Order.fromJson(body);
    } else {
      throw Exception('Failed to get placeOrder response');
    }
  }

}

File: examples/online-boutique/backend/services/checkout/checkout.go
//ftl:module checkout
package checkout

import (
	"context"
	"fmt"

	"github.com/google/uuid"

	"ftl/currency"
	"ftl/payment"
	"ftl/productcatalog"
	"ftl/shipping"

	"ftl/builtin"
	"ftl/cart"

	"github.com/TBD54566975/ftl/examples/online-boutique/backend/common/money"

	"github.com/TBD54566975/ftl/go-runtime/ftl"
)

type PlaceOrderRequest struct {
	UserCurrency string

	Address    shipping.Address
	Email      string
	CreditCard payment.CreditCardInfo
}

type OrderItem struct {
	Item cart.Item
	Cost currency.Money
}

type Order struct {
	ID                 string
	ShippingTrackingID string
	ShippingCost       currency.Money
	ShippingAddress    shipping.Address
	Items              []OrderItem
}

type ErrorResponse struct {
	Message string `json:"message"`
}

//ftl:ingress POST /checkout/{userId}
func PlaceOrder(ctx context.Context, req builtin.HttpRequest[PlaceOrderRequest, string, ftl.Unit],
	getCartClient cart.GetCartClient,
	productClient productcatalog.GetClient,
	currencyConverter currency.ConvertClient,
	shippingQuote shipping.GetQuoteClient,
	chargeClient payment.ChargeClient,
	shipper shipping.ShipOrderClient,
	emptyCart cart.EmptyCartClient) (builtin.HttpResponse[Order, ErrorResponse], error) {
	logger := ftl.LoggerFromContext(ctx)

	cartItemsResp, err := getCartClient(ctx, builtin.HttpRequest[ftl.Unit, ftl.Unit, cart.GetCartRequest]{Query: cart.GetCartRequest{UserId: req.PathParameters}})
	if err != nil {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to get cart for user %q: %s", req.PathParameters, err)}),
		}, nil
	}

	cartItems, ok := cartItemsResp.Body.Get()
	if !ok {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to get cart for user %q %s", req.PathParameters, cartItemsResp.Error.MustGet())}),
		}, nil
	}

	orders := make([]OrderItem, len(cartItems.Items))
	for i, item := range cartItems.Items {
		productsResp, err := productClient(ctx, builtin.HttpRequest[ftl.Unit, productcatalog.GetRequest, ftl.Unit]{PathParameters: productcatalog.GetRequest{Id: item.ProductId}})
		if err != nil {
			return builtin.HttpResponse[Order, ErrorResponse]{
				Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to get product #%q: %s", item.ProductId, err)}),
			}, nil
		}

		products, ok := productsResp.Body.Get()
		if !ok {
			return builtin.HttpResponse[Order, ErrorResponse]{
				Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("product not found: %q %s", item.ProductId, productsResp.Error.MustGet())}),
			}, nil
		}

		priceResp, err := currencyConverter(ctx, builtin.HttpRequest[currency.ConvertRequest, ftl.Unit, ftl.Unit]{
			Body: currency.ConvertRequest{
				From:   products.PriceUsd,
				ToCode: req.Body.UserCurrency,
			},
		})
		if err != nil {
			return builtin.HttpResponse[Order, ErrorResponse]{
				Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to convert price of %q to %s: %s", item.ProductId, req.Body.UserCurrency, err)}),
			}, nil
		}

		price, ok := priceResp.Body.Get()
		if !ok {
			return builtin.HttpResponse[Order, ErrorResponse]{
				Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to convert price of %q to %s %s", item.ProductId, req.Body.UserCurrency, priceResp.Error.MustGet())}),
			}, nil
		}

		orders[i] = OrderItem{
			Item: item,
			Cost: price,
		}
	}

	shippingUSDResp, err := shippingQuote(ctx, builtin.HttpRequest[shipping.ShippingRequest, ftl.Unit, ftl.Unit]{
		Body: shipping.ShippingRequest{
			Address: req.Body.Address,
			Items:   Map(orders, func(i OrderItem) cart.Item { return i.Item }),
		},
	})
	if err != nil {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to get shipping quote: %s", err)}),
		}, nil
	}

	shippingUSD, ok := shippingUSDResp.Body.Get()
	if !ok {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to get shipping quote %s", shippingUSDResp.Error.MustGet())}),
		}, nil
	}

	shippingPriceResp, err := currencyConverter(ctx, builtin.HttpRequest[currency.ConvertRequest, ftl.Unit, ftl.Unit]{
		Body: currency.ConvertRequest{
			From:   shippingUSD,
			ToCode: req.Body.UserCurrency,
		},
	})
	if err != nil {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to convert shipping cost to currency: %v", err)}),
		}, nil
	}

	shippingPrice, ok := shippingPriceResp.Body.Get()
	if !ok {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to convert shipping cost to currency %s", shippingPriceResp.Error.MustGet())}),
		}, nil
	}

	total := currency.Money{CurrencyCode: req.Body.UserCurrency}
	total = money.Must(money.Sum(total, shippingPrice))
	for _, it := range orders {
		multPrice := money.MultiplySlow(it.Cost, uint32(it.Item.Quantity))
		total = money.Must(money.Sum(total, multPrice))
	}
	txIDResp, err := chargeClient(ctx, builtin.HttpRequest[payment.ChargeRequest, ftl.Unit, ftl.Unit]{
		Body: payment.ChargeRequest{
			Amount:     total,
			CreditCard: req.Body.CreditCard,
		},
	})
	if err != nil {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{
				Message: fmt.Sprintf("failed to charge card: %s", err),
			}),
		}, nil
	}

	txID, ok := txIDResp.Body.Get()
	if !ok {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to charge card %v", txIDResp.Error.MustGet())}),
		}, nil
	}

	logger.Infof("Charged card, ID %s", txID.TransactionId)

	shippingTrackingIDResp, err := shipper(ctx, builtin.HttpRequest[shipping.ShippingRequest, ftl.Unit, ftl.Unit]{
		Body: shipping.ShippingRequest{
			Address: req.Body.Address,
			Items:   cartItems.Items,
		},
	})
	if err != nil {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("shipping error: %s", err)}),
		}, nil
	}

	shippingTrackingID, ok := shippingTrackingIDResp.Body.Get()
	if !ok {
		return builtin.HttpResponse[Order, ErrorResponse]{
			Error: ftl.Some(ErrorResponse{Message: fmt.Sprintf("failed to ship order %s", shippingTrackingIDResp.Error.MustGet())}),
		}, nil
	}

	logger.Infof("Shipped order, tracking ID %s", shippingTrackingID.Id)

	// Empty the cart, but don't worry about errors.
	_, _ = emptyCart(ctx, builtin.HttpRequest[cart.EmptyCartRequest, ftl.Unit, ftl.Unit]{Body: cart.EmptyCartRequest{UserId: req.PathParameters}})

	order := Order{
		ID:                 uuid.New().String(),
		ShippingTrackingID: shippingTrackingID.Id,
		ShippingCost:       shippingPrice,
		ShippingAddress:    req.Body.Address,
		Items:              orders,
	}
	// if err := s.emailService.Get().SendOrderConfirmation(ctx, req.Email, order); err != nil {
	// 	s.Logger(ctx).Error("failed to send order confirmation", "err", err, "email", req.Email)
	// } else {
	// 	s.Logger(ctx).Info("order confirmation email sent", "email", req.Email)
	// }

	return builtin.HttpResponse[Order, ErrorResponse]{Body: ftl.Some(order)}, nil
}

func Map[T, U any](slice []T, fn func(T) U) []U {
	result := make([]U, len(slice))
	for i, v := range slice {
		result[i] = fn(v)
	}
	return result
}
